notebook-api/notebook/notebook.py

```diff
@@ -0,0 +1,434 @@
+        for path in paths:
+            if len(path) == 3:
+                data[path[0]] = path[2](get_path(rsc, path[1]))
+            else:
+                data[path[0]] = get_path(rsc, path[1])
+
+        resp.append(data)
+
+    if flatten == True and len(resources) == 1:
+        return ujson.dumps(resp[0])
+
+    return ujson.dumps(resp)
+
+
+# Services and pods are seperate resources, and operations between them are not atomic
+# Must support deletion of one indep of presence other
+
+
+def del_pod(pod_name):
+    k8s.delete_namespaced_pod(
+        pod_name,
+        'default',
+        kube.client.V1DeleteOptions(),
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+
+
+def del_svc(svc_name):
+    k8s.delete_namespaced_service(
+        svc_name,
+        'default',
+        kube.client.V1DeleteOptions(),
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+
+
+##### New notebook methods meant for consumption from agnostic client#####
+###### Methods that do not call the auth-gateway /verify method must not be directly public #######
+
+# TODO: learn how to properly handle webscocket close in gevent + wsgi
+# or just move to aiohttp and stop dealing with greenlets and websockets with
+# no bound events (though there may be a way in gevent's websocket package + wsgi)
+# simply checkingfo ws.close isn't enough
+# https://github.com/heroku-python/flask-sockets/issues/60
+# A reactive approach to websockets. Primary benefit is 1 connection per user
+# rather than 1 connection per pod
+# no need to go to public web to hit http endpoint
+# and real insight into pod status
+# Weakness is currently not checking whether svc is accessible; easily can add
+
+
+@sockets.route('/api/ws')
+def echo_socket(ws):
+    user_id = ws.environ['HTTP_USER']  # and scope is ws.environ['HTTP_SCOPE']
+
+    k8s = kube.client.CoreV1Api()
+
+    w = kube.watch.Watch()
+
+    for event in w.stream(k8s.list_namespaced_pod, namespace='default', watch=False,
+                          label_selector=f"user_id={UNSAFE_user_id_transform(user_id)}"):
+
+        # This won't prevent socket is dead errors, presumably something related to
+        # greenlet socket handling and our inability to add an on_closed callback to end watch
+        if ws.closed:
+            return
+
+        try:
+            obj = event["object"]
+            print("Event: %s %s %s" %
+                  (event['type'], event['object'].kind, event['object'].metadata.name))
+
+            if event['type'] == 'MODIFIED' or event['type'] == 'ADDED':
+                ws.send(
+                    f'{{"event": "{event["type"]}", "resource": {marshall_json([obj], pod_paths, True)}}}')
+
+            if event['type'] == 'DELETED':
+
+                ws.send(
+                    f'{{"event": "DELETED", "resource": {marshall_json([obj], pod_paths, True)}}}')
+        except Exception as e:
+            log.error(e)
+            return
+
+    while not ws.closed:
+        gevent.sleep(10)
```

Comment:
This was an early attempt at checking for ws.closed status. The problem is the advertised approach with WSGI does not give us a chance to watch for ws closed events...which is silly and unfortunate (there's hopefully a way to do this, if not there def is in the asyncio lib), as it prevents us from correctly, explicitly tearing down resources that depend on it.

I now try to explicitly close the kube connection, and as early as possible, because I'm not confident gevent is doing the right thing.